### PR TITLE
fix(amazonqFeatureDev): mitigate race conditions on session creation

### PR DESCRIPTION
**Pull request by sannicm**
## Problem

When a customer starts a /dev chat with a message, e.g. /dev <do something>, the connector logic calls two event handlers in amazonqFeatureDev logic parallely that try to create a session, this creates a race condition where the customer could get to code generation with the wrong session, i.e. one with no conversationId.

## Solution

Add a lock to the getSession function in order to mitigate concurrent calls to getSession.

This is a similar approach to the one taken in JetBrain's toolkit: [ChatSessionStorage.kt](https://github.com/aws/aws-toolkit-jetbrains/blob/ec168fc6e9306f22f07945fd516d6acdc7fe396e/plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/storage/ChatSessionStorage.kt#L18)

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by jpinkney-aws — APPROVED** [review]
@sannicm Just to link everything together can you point a link to the jetbrains code where we had to do the same thing

**Comment by justinmk3** [comment]
Looks like yet another reason why the "session" concept shouldn't be copy-pasted in each Q subfeature. Will other Q features (transform, refactor-assist, etc) potentially encounter the same bug?

Is there a customer-facing improvement that can be mentioned in the changelog? E.g. "Amazon Q [subfeature]: potential duplicate sessions"

File: packages/core/src/amazonqFeatureDev/storages/chatSession.ts [hunk 1]
```diff
@@ -20,12 +23,14 @@ export class ChatSessionStorage {
     }
 
     public async getSession(tabID: string): Promise<Session> {
-        const sessionFromStorage = this.sessions.get(tabID)
-        if (sessionFromStorage === undefined) {
-            // If a session doesn't already exist just create it
-            return this.createSession(tabID)
-        }
-        return sessionFromStorage
+        return this.lock.acquire(tabID, async () => {
```
Comment (ivikash) [inline packages/core/src/amazonqFeatureDev/storages/chatSession.ts:26]: I'm curious if we can have `this.lock.acquire` as decorator?

**Comment by sannicm** [comment]
> @sannicm Just to link everything together can you point a link to the jetbrains code where we had to do the same thing

https://github.com/aws/aws-toolkit-jetbrains/blob/main/plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonqFeatureDev/storage/ChatSessionStorage.kt#L18

**Comment by sannicm** [comment]
> Looks like yet another reason why the "session" concept shouldn't be copy-pasted in each Q subfeature. Will other Q features (transform, refactor-assist, etc) potentially encounter the same bug?
> 
> Is there a customer-facing improvement that can be mentioned in the changelog? E.g. "Amazon Q [subfeature]: potential duplicate sessions"

I cannot estimate the chances they will encounter the same issue as every feature has its own integration with mynah-ui. It depends if their implementation of the integration with mynah-ui/amazonq chat fires multiple events together and if they handle that correctly.

For amazonqFeatureDev, the best solution would be to update the `handleDevFeatureCommand` in the quick actions: `src/amazonq/webview/ui/quickActions/handler.ts` file making it only send one event (e.g. just using the `onKnownTabOpened`) even if there is a command message, i.e. the text in `/dev <text>` which currently we do not have capacity, it will require a parallel effort in JetBrains to update the same and test. 

This is not though reducing the code duplication you notice.

File: packages/core/src/amazonqFeatureDev/storages/chatSession.ts (hunk 1 above)
Comment (sannicm) [reply packages/core/src/amazonqFeatureDev/storages/chatSession.ts:31]: That is still a typescript's experimental feature, correct?

Comment (justinmk3) [reply packages/core/src/amazonqFeatureDev/storages/chatSession.ts:31]: Would that make the code clearer, or less obvious?

**Pull request merged**